website/integrations: add hass-openid instructions (#14672)
goauthentik/authentik: website/integrations/services/home-assistant/index.md

This documentation lists only the settings that you need to change from their default values. Be aware that any changes other than those explicitly mentioned in this guide could cause issues accessing your application.
:::

> ## Using OAuth2 / OpenID

Reviewer: ```suggestion

Home Assistant can use either [OAuth2 / OpenID](#Using OAuth2 / OpenID) or a [proxy provider](#Using a proxy provider). See the relevant instructions for each below.

## Using OAuth2 / OpenID
```

> ### authentik configuration

Reviewer: Can we say OIDC or does it need to be OpenID? @cavefire

2. Navigate to **Applications** > **Applications** and click **Create with Provider** to create an application and provider pair. (Alternatively you can first create a provider separately, then create the application and connect it with the provider.)

    - **Application**: provide a descriptive name, an optional group for the type of application, the policy engine mode, and optional UI settings.
    - **Choose a Provider type**: select **OAuth2/OpenID** as the provider type.
    
    - Note the **Client ID**,**Client Secret**, and **slug** values because they will be required later.
        - **Signing Key**: Select any available signing key.
        - **Client ID**: Use the value generated by authentik.
        - **Client Secret**: Use the value generated by authentik.
        - **Redirect URIs**:
            - Regex: <kbd>^(http:\/\/hass\.company\:8123)\/auth\/openid\/callback.*</kbd>

    - **Configure Bindings** _(optional)_: you can create a [binding](/docs/add-secure-apps/flows-stages/bindings/) (policy, group, or user) to manage the listing and access to applications on a user's **My applications** page.

3. Click **Submit** to save the new application and provider.

> ### Home Assistant configuration

Reviewer: ```suggestion
### Home Assistant configuration for OAuth2 / OpenID
```

1. Follow the installation guide on [https://github.com/cavefire/hass-openid](https://github.com/cavefire/hass-openid?tab=readme-ov-file#installation).
2. Using a file editor or ssh, edit the file `configuration.yaml` in your homeassistant's `config` directory by appending the following:
    ```yaml
    openid:
        client_id: <Client ID from Step 1>
        client_secret: <Client Secret from Step 1>
        authorize_url: "https://authentik.company/application/o/authorize/"
        token_url: "https://authentik.company/application/o/token/"
        user_info_url: "https://authentik.company/application/o/userinfo/"
        scope: "openid profile email"
        username_field: "preferred_username"
    ```
3. Restart Home Assistant

The login page now has a button called "OpenID/OAuth2 authentication". Click it to redirect you to authentik to sign in.
Make sure the user account has been created in Home Assistant, because this integration does not create the users for you.

> ## Using a proxy

Reviewer: ```suggestion
## Using a proxy provider
```

    - **Application**: provide a descriptive name, an optional group for the type of application, the policy engine mode, and optional UI settings.
    - **Choose a Provider type**: select **Proxy** as the provider type.
    
      - **External Host**: Set this to the external URL you will be accessing Home Assistant from
      - **Client ID**: Use the value generated by authentik.
      - **Internal Host**: <kdb>http://hass.company:8123</kdb>

    - **Configure Bindings** _(optional)_: you can create a [binding](/docs/add-secure-apps/flows-stages/bindings/) (policy, group, or user) to manage the listing and access to applications on a user's **My applications** page.

3. Click **Submit** to save the new application and provider.

4. Create an outpost deployment for the provider you've created above, as described [here](https://docs.goauthentik.io/docs/add-secure-apps/outposts/). Deploy this Outpost either on the same host or a different host that can access Home Assistant. The outpost will connect to authentik and configure itself.

> ### Home Assistant configuration

Reviewer: ```suggestion
### Home Assistant configuration with a proxy provider
```

This documentation lists only the settings that you need to change from their default values. Be aware that any changes other than those explicitly mentioned in this guide could cause issues accessing your application.
:::

> Home Assistant can use either [OIDC](#using-OIDC) or a [proxy-provider](#Using a proxy provider). See the relevant instructions for each below.

Reviewer 2: Can you please re-check these anchors? For example, they can't contain spaces

## Using OIDC

### authentik configuration for OIDC

To support the integration of `Home Assistant` using `hass-openid` with authentik, you need to create an application/provider pair in authentik.

#### Create an application and provider in authentik

1. Log in to authentik as an admin, and open the authentik Admin interface.
2. Navigate to **Applications** > **Applications** and click **Create with Provider** to create an application and provider pair. (Alternatively you can first create a provider separately, then create the application and connect it with the provider.)

    - **Application**: provide a descriptive name, an optional group for the type of application, the policy engine mode, and optional UI settings.
    - **Choose a Provider type**: select **OAuth2/OpenID** as the provider type.
    
    - Note the **Client ID**,**Client Secret**, and **slug** values because they will be required later.
        - **Signing Key**: Select any available signing key.
        - **Client ID**: Use the value generated by authentik.
        - **Client Secret**: Use the value generated by authentik.
        - **Redirect URIs**:

>             - Regex: <kbd>^(http:\/\/hass\.company\:8123)\/auth\/openid\/callback.*</kbd>

Reviewer 2: I've mentioned it in a later comment, so this too
```suggestion
            - Regex: `^(https:\/\/hass\.company)\/auth\/openid\/callback.*`
```

    - **Configure Bindings** _(optional)_: you can create a [binding](/docs/add-secure-apps/flows-stages/bindings/) (policy, group, or user) to manage the listing and access to applications on a user's **My applications** page.

3. Click **Submit** to save the new application and provider.

### Home Assistant configuration for OIDC

> 1. Follow the installation guide on [https://github.com/cavefire/hass-openid](https://github.com/cavefire/hass-openid?tab=readme-ov-file#installation).

Author: I don't want to add the documentation here, since this would result in the documentation being in two places, potentially being out of sync confusing the user.

> 1. Click the following link to install the custom component: [Install hass-openid using HACS](https://my.home-assistant.io/redirect/hacs_repository/?category=integration&repository=hass-openid&owner=cavefire)

Reviewer 3: ```suggestion
1. Install hass-openid following the instructions at https://github.com/cavefire/hass-openid
```

> 2. Using a file editor or ssh, edit the file `configuration.yaml` in your homeassistant's `config` directory by appending the following:

Reviewer: ```suggestion
2. Using a file editor or ssh, edit the file `configuration.yaml` in your Home Assistant's `config` directory by appending the following:
```

        username_field: "preferred_username"
    ```
3. Restart Home Assistant

The login page now has a button called "OpenID/OAuth2 authentication". Click it to redirect you to authentik to sign in.
Make sure the user account has been created in Home Assistant, because this integration does not create the users for you.

## Using a proxy provider

:::caution
Using a proxy provider might produce CSRF errors; this is caused by a technology that Home Assistant uses and not authentik. For more information see [this GitHub issue](https://github.com/goauthentik/authentik/issues/884#issuecomment-851542477).
:::
:::caution
Only prefixes starting with `/auth` need to be proxied (excluding prefixes starting with `/auth/token`). See [this GitHub issue](https://github.com/BeryJu/hass-auth-header/issues/212). This can be configured in the reverse proxy (e.g. nginx, Traefik) or in authentik Provider's **Unauthorized Paths**.
:::

### authentik configuration with a proxy provider

To support the integration of `Home Assistant` using `hass-auth-headers` with authentik, you need to create an application/provider pair in authentik.

#### Create an application and provider in authentik

> 1. Log in to authentik as an admin, and open the authentik Admin interface.

Reviewer 2: ```suggestion
1. Log in to authentik as an administrator, and open the authentik Admin interface.
```

2. Navigate to **Applications** > **Applications** and click **Create with Provider** to create an application and provider pair. (Alternatively you can first create a provider separately, then create the application and connect it with the provider.)

    - **Application**: provide a descriptive name, an optional group for the type of application, the policy engine mode, and optional UI settings.
    - **Choose a Provider type**: select **Proxy** as the provider type.

>     

Reviewer 2: ```suggestion
```

      - **External Host**: Set this to the external URL you will be accessing Home Assistant from
      - **Client ID**: Use the value generated by authentik.

>       - **Internal Host**: <kdb>http://hass.company:8123</kdb>

Reviewer 2: ```suggestion
      - **Internal Host**: `http://hass.company:8123`
```

Can't you run Home assistant on :443 w/ https as well? Generally, we prefer using:
```suggestion
      - **Internal Host**: `https://hass.company`
```
and users can update the URL as needed

Author: Home Assistant does not support ssl by default. Therefore using the default port 8123 for the user to easily understand.

This documentation lists only the settings that you need to change from their default values. Be aware that any changes other than those explicitly mentioned in this guide could cause issues accessing your application.
:::

## Using OAuth2 / OpenID

### authentik configuration

> 1. Create a **OAuth2/OpenID Provider** under **Aplications** > **Providers**:

Reviewer: Hi @cavefire and thanks so much for this PR. May I ask for a modification to the "authentik configuration" section, please? A few releases ago we provided the ability to create the app/provider pair together, as a single process. Our integration template [service.md](https://github.com/goauthentik/authentik/blob/main/website/integrations/template/service.md) in our GitHub repo reflects this change, and the proposed wording for using the new process. Or let me know if you prefer I make the change and push back to your PR. Thanks again!

Author: Done that. @tanberry, can you please check if that fits your style?

Author: I did the rebase.

The repetition is intended. Home Assistant does not support OAuth2 out of the box. Because of that there is a custom integration, adding compatibility using the proxy provider: https://github.com/BeryJu/hass-auth-header
That was described previously.

Now I have added a new section for OAuth2, because I have created a new Home Assistant integration, that does not need a proxy, instead using OAuth2. I haven't remove the proxy approach from the documentation, rather just added the OAuth2 approach on top. Thats why the repetition is present.

Reviewer: Also @cavefire let me know if you want to rebase and try to get the build errors fixed, or if you want me to (after you make any remaining edits).

Author: I replaced everything with OIDC. Thanks for your feedback and help!

    - Note the **Client ID** and **Client Secret** values because they will be required later.
    - Set a `Regex` redirect URI to `^(http:\/\/hass\.company\:8123)\/auth\/openid\/callback.*`. (Note the `\` in front of every `.`, `/` and `:`. Check you regex [here](https://regex101.com)
    - Select any available signing key.
- **Configure Bindings** _(optional)_: you can create a [binding](/docs/add-secure-apps/flows-stages/bindings/) (policy, group, or user) to manage the listing and access to applications on a user's **My applications** page.

2. Create an **Application** under **Applications** > **Applications** using the following settings:

    - **Name**: Home Assistant
    - **Slug**: homeassistant
    - **Provider**: Home Assistant (the provider you created in step 1)

### Home Assistant configuration

1. Follow the installation guide on [https://github.com/cavefire/hass-openid](https://github.com/cavefire/hass-openid?tab=readme-ov-file#installation)

> 2. Using a file editor or ssh, edit the file `configuration.yaml` in your homeassistant's `config` by appending the following:

Reviewer: @cavefire is it indeed a file, with the name `config`? Let's add the extension too, please.

> 2. Using a file editor or ssh, edit the file `configuration.yaml` in your homeassistant's `config` by appending the following:

Author: @tanberry Its a directory. I changed it accordingly.
Thanks!

    ```yaml
    openid:
        client_id: <Client ID from Step 1>
        client_secret: <Client Secret from Step 1>
        authorize_url: "https://authentik.company/application/o/authorize/"
        token_url: "https://authentik.company/application/o/token/"
        user_info_url: "https://authentik.company/application/o/userinfo/"
        scope: "openid profile email"
        username_field: "preferred_username"
    ```

> 3. Restart Homeassistant

Reviewer: ```suggestion
3. Restart Home Assistant
```

> The login page now has a button called "OpenID/OAuth2 authentication". Clicking it redirects you to authentik to sign in.

Reviewer: ```suggestion
The login page now has a button called "OpenID/OAuth2 authentication". Click it to redirect you to authentik to sign in.
```

Make sure the user account has been created in Homeassistant, since this integration does not create the users for you.

## Using a proxy

:::caution

> You might run into CSRF errors, this is caused by a technology Home-assistant uses and not authentik, see [this GitHub issue](https://github.com/goauthentik/authentik/issues/884#issuecomment-851542477).

Reviewer: ```suggestion
Using a proxy provider might produce CSRF errors; this is caused by a technology that Home Assistant uses and not authentik. For more information see [this GitHub issue](https://github.com/goauthentik/authentik/issues/884#issuecomment-851542477).
```

:::
:::caution

> Only prefixes starting with `/auth` need to be proxied (excluding prefixes starting with `/auth/token`), see [this GitHub issue](https://github.com/BeryJu/hass-auth-header/issues/212). This can be configured in the reverse proxy (e.g. nginx, Traefik) or in authentik Provider's **Unauthorized Paths**.

Reviewer: ```suggestion
Only prefixes starting with `/auth` need to be proxied (excluding prefixes starting with `/auth/token`). See [this GitHub issue](https://github.com/BeryJu/hass-auth-header/issues/212). This can be configured in the reverse proxy (e.g. nginx, Traefik) or in authentik Provider's **Unauthorized Paths**.
```
